### PR TITLE
Add [@poll error] attribute

### PR DESCRIPTION
**Pull request by sadiqj**
**Summary**: This PR proposes an attribute to accompany the safepoints PR (https://github.com/ocaml/ocaml/pull/10039) that will help developers of libraries containing atomic sections to do so in a less error prone way. It is stacked atop of the safepoints PR and the commit of interest for the attribute is https://github.com/ocaml/ocaml/commit/205658d88eaa46c5fabe1eb69422e58bf3c104b2. 

----

The safepoints PR (https://github.com/ocaml/ocaml/pull/10039) introduces polling operations to code generated by the native compiler. These use the same interrupt checking mechanism as allocation, triggering if the minor allocation pointer crosses the minor allocation limit. Safepoints fixes some long-standing issues around delivery of signals (https://github.com/ocaml/ocaml/issues/3747) and correctness (https://github.com/ocaml/ocaml/issues/7128). Frequent polling is also necessary for inter-thread garbage collection coordination in Multicore.

Since a poll operation might lead to a garbage collection and this might involve running arbitrary OCaml code through finalisers, it presents an issue for users who wish to write programs with atomic sections. At present avoiding allocations, which itself can be tricky, is all that is required.

The safepoints implementation gives clear rules to the user for where polling operations may appear:
- In function prologues
- At the back edge of loops

In order to make writing atomic sequences less error prone, an attribute which would lead to a compiler error if a user inadvertently introduces a polling operation would be desirable.

The proposal is for a function attribute which would result in a compiler error if the compiled function would have any of the following:

- Allocations
- Polling operations
- Calls to OCaml functions
- Calls to non-noalloc C functions

In addition, a function marked with this attribute would mean the function was non-inlinable and is also ‘never’ for the local function optimisation. This is necessary so we can track this restriction at the level of whole functions through the compiler. Enforcing the restriction at the block level through the middle and back ends would be a complex endeavour.

With this PR the following:

```
let[@inline never][@local never] v x = x + 1

let[@poll error] c x =
  let y = Sys.opaque_identity(ref 42) in
    let x2 = v x in
      for c = 0 to x2 do
        ignore(Sys.opaque_identity(42))
      done;
      x2 + !y
````

gives a compile error of:

````
File "foo.ml", line 1:
Error: Polling instructions in function annotated with [@poll error]
        inserted poll at File "_none_", line 1
        function call at File "foo.ml", line 5, characters 13-16
        allocation at File "foo.ml", line 4, characters 29-37
````

(The location information for the poll needs work)

Use of this attribute exposes the user to changes in the compiler implementation. For example, a change to the polling eliding algorithm could introduce polls and thus cause code to no longer compile. The same applies to changes in inlining heuristics and optimisation passes that eliminate allocations. We will need to highlight this in documentation.

The commit containing the attribute is https://github.com/ocaml/ocaml/commit/205658d88eaa46c5fabe1eb69422e58bf3c104b2 , the previous commits are from the safepoints PR itself. There is currently no support for Flambda but this approach has been discussed with @mshinwell and once there is consensus on the design, we will add support.

### CODE REVIEW

**Comment by gadmm** [comment]
Thanks, this is a highly-awaited feature to enforce reasoning about polling locations. (I think this also interests @stedolan and @lthls.) It would be nice to see examples of applications (e.g. how do you rewrite Nano_mutex? Do you lose performance due to lack of inlining of the new atomic sections?)

I think the current proposal fails to enable the reasoning about polling locations used for instance here in Coq: <https://github.com/coq/coq/blob/271445decd0fc1a37da3009f148f2e68c7168fe1/clib/cThread.ml#L111-L128>. It is a `with_lock` combinator that is protected against asynchronous exceptions. Without it, Coq is buggy. Let me reproduce the code here:
```ocaml
external unlock: t -> unit = "caml_mutex_unlock"

(* Critical sections are enforced because:
   - Mutex.lock does not poll on leaving the blocking section
     since 4.12.
   - Never inline, to avoid theoretically-possible reorderings with
     flambda.
   - We inline the call to Mutex.unlock to avoid polling in bytecode
     mode.
*)
let[@inline never] with_lock m ~scope =
  let () = lock m (* BEGIN ATOMIC *) in
  match (* END ATOMIC *) scope () with
  | (* BEGIN ATOMIC *) x -> unlock m ; (* END ATOMIC *) x
  | (* BEGIN ATOMIC *) exception e -> unlock m ; (* END ATOMIC *) raise e
```
I like this example because it is very similar to what is needed for an async-safe bracket (https://github.com/ocaml/ocaml/pull/8962, [code in question](https://github.com/ocaml/ocaml/pull/8962/commits/5bec330b37dfc03c851ec30be38dd22b0b21c335#diff-3281003824dbd4299b844117c245dab1aff4aa02452d0c534fd096d388105ad8R48-R87)) but without the masking that requires changes to the runtime. In other words it is current OCaml, while being representative of much more.

I tried it with your branch here: https://github.com/gadmm/ocaml/tree/poll_error_attribute. After replacing `[@inline never]` with `[@poll error]` I get the following errors:
```
../../ocamlopt.opt -nostdlib -I ../../stdlib -I ../../otherlibs/unix -c -w +33..39 -warn-error +A -g -bin-annot -safe-string -function-sections mutex.ml
File "mutex.ml", line 1:
Error: Polling instructions in function annotated with [@poll error]
	alloc external call at File "mutex.ml", line 35, characters 28-36
	alloc external call at File "mutex.ml", line 36, characters 38-46
	function call at File "mutex.ml", line 34, characters 25-33
	alloc external call at File "mutex.ml", line 33, characters 11-17
```
This complains that `lock` and `scope` can allocate, but those polling locations are fine because they are outside of the atomic sections. In addition, it says that `unlock` can poll. Unfortunately this is an approximation: it is a non-noalloc C call, but it does not allocate or poll inside the atomic section (it cannot be `noalloc` because it can raise in other situations involving bugs which does not concern us here).

In other words, the current proposal is too coarse for this real-world situation. What's more, can we really find a way to express in the language these `BEGIN/END ATOMIC` sections? They are not even blocks, but work across block boundaries (including, in the case of `lock`, starting in the middle of a C function).

So, we have to make a difference between the fine-grained atomic sections and the coarse-grained attribute. We first need to clarify the purpose of the attribute from this point of view: it is not to delineate atomic sections, but to help/enable the programmer to reason about such finer-grained atomic sections contained in the annotated function (which themselves might have to remain documented as `BEGIN/END ATOMIC` comments).

What we need here from such an attribute to reason about polling locations is:
- to ensure that the user can easily predict where polling happens, in particular that no additional polling location is inserted (the Safepoints PR mentions adding a polling location when non-noalloc C calls return, which should be turned off if this happened, and in the case of #8962 which uses sub-functions, no polling should happen in their prologue either for functions under this attribute).
- to ensure that flambda adopts a strict semantics regarding polling locations when transforming programs (to avoid the issue described at #10035).
- to ensure that the bytecode behaves the same as native on this code (here only the inlining of `unlock` was needed, whereas #8962 uses a weird twist to work in bytecode).

Can we make a coarse-grained attribute work for whole functions only instead of blocks, for simplicity of implementation as you propose? As far as I can tell from practical experiments, yes, this would be enough (this is after working hard to simplify the implementation of resource combinators to see what kind of annotation could realistically work; some of you may remember private discussions where I was looking at much more complicated semantics).

The attribute can have the following meaning:
1. assert the absence of non-explicit polling location.
2. disable the insertion of polling locations.
3. tell the compiler that polling locations should be interpreted strictly (in bytecode and by flambda).

For 1., you propose to make it an error if there is any sort of polling location, with some conservative approximation. In the example above, this is too restrictive. Instead, I propose to allow calls to OCaml functions and non-noalloc C functions (on the principle that the programmer can easily spot these and reason about them). Implicit allocations (such as when boxing an unboxed float argument) should cause an error (and actually a very conservative approximation would be enough and probably better than something that might change over time, but as I understand your implementation is simpler since it directly checks the back-end). As for explicit allocations (e.g. allocating a pair), they can remain forbidden: in my use-cases cases, I managed to remove all explicit allocations, and if I really needed I could wrap the allocation in a function (since those would be allowed). According to the error messages you display, you already have a distinction between calls and other kinds of allocations so I expect that allowing calls would be easy to implement.

For 2., you propose to make it an error if any polling location is inserted. Instead I propose to disable the insertion of polling locations. The user is then expected to make sure there are enough polling locations for multicore synchronisation (in my use-cases this is not an issue since the control flow is flat). It is probably easy to turn off the insertion of all polling locations on a per-whole-functions basis.

For 3., you do not describe what is needed for flambda support, and flambda is not well documented so it is hard for me to discuss it. Regarding the problem I am aware of (#10035), it might be enough to forbid all transformations and inlining in a first time. As for bytecode, support can be implemented later separately: I think a requirement for this proposal is instead that it has a clear meaning for bytecode so that it can be done later, maybe by someone else, whenever necessary (for instance when I need it in my work on resource management, but I prefer to make no commitment at this point).

To sum up:

| | Your proposal | My proposal |
| --- | --- | --- |
| Explicit allocation | Error | Error or Allow |
| Implicit allocation | Error | Error |
| Call OCaml function | Error | Allow |
| Call C function | Error if non-noalloc | Allow |
| Inserted polls | Error | Disable |

I hope you will like my proposal, and that it is as easy to implement as I think.

File: asmcomp/polling.ml [hunk 1]
```diff
@@ -178,18 +183,69 @@ let instr_body handler_safe i =
...
-  let new_contains_calls = f.fun_contains_calls || contains_poll new_body in
-  { f with fun_body = new_body; fun_contains_calls = new_contains_calls }
+  if f.fun_poll_error then begin
+    let poll_instrs = contains_poll_alloc_or_calls new_body in
+    match poll_instrs with
+    | [] -> ()
+    | _ -> raise (Error (Poll_error poll_instrs))
+  end;
+    let new_contains_calls = f.fun_contains_calls || contains_poll new_body in
+    { f with fun_body = new_body; fun_contains_calls = new_contains_calls }
```
Comment (gadmm) [inline asmcomp/polling.ml:289]: Indentation is strange here (and elsewhere but this is an example where it affects readability).

File: lambda/translattribute.ml [hunk 2]
```diff
@@ -192,6 +213,23 @@ let check_local_inline loc attr =
   | _ ->
       ()
 
+let check_poll_inline loc attr =
+  match attr.poll, attr.inline with
+  | Error_poll, (Always_inline | Hint_inline | Unroll _) ->
+      Location.prerr_warning loc
+        (Warnings.Inlining_impossible "[@error_poll] disables inlining")
```
Comment (gadmm) [inline lambda/translattribute.ml]:
```suggestion
        (Warnings.Inlining_impossible "[@poll error] is incompatible with inlining")
```

File: lambda/translattribute.ml [hunk 3]
```diff
@@ -192,6 +213,23 @@ let check_local_inline loc attr =
...
+  | _ ->
+      ()
+
+let check_poll_local loc attr =
+  match attr.poll, attr.local with
+  | Error_poll, Always_local ->
+      Location.prerr_warning loc
+        (Warnings.Inlining_impossible
+          "[@error_poll] disables local function optimization")
```
Comment (gadmm) [inline lambda/translattribute.ml]:
```suggestion
          "[@poll error] is incompatible with local function optimization")
```

**Review by gadmm — COMMENTED** [review]
Some cosmetic remarks

**Comment by gadmm** [comment]
I attempted to implement my proposal at https://github.com/gadmm/ocaml/tree/poll_error_attribute with a new `[@poll explicit]` attribute. There is very little change. But it would need to implement one more thing: prevent the inlining of any contained function call. Indeed, the following two (bogus) functions `g` should be different:
```ocaml
let rec f x = f x

let[@poll explicit] g x = f x
```
```ocaml
let[@poll explicit] g x = 
  let rec f x = f x in
  f x
```
Only the latter should cause other domains to deadlock. The disabling of poll insertion should not propagate across function calls via inlining. (Idem for erroring on allocation when you define `let f x = (x,x)`.)

**Comment by dra27** [comment]
I'm not convinced that we should be providing attributes like `[@poll explicit]` which _control_ the backend in this way. The ability to suppress poll generation sounds like a debugging nightmare when some library incorrectly flags functions as atomic because a naive programmer vaguely read about it and tagged some networking operations as needing to be atomic.

Other pure OCaml examples I understand will be forthcoming. For your example from Coq, if we take my assertion that I don't think `@poll explicit` should be permitted, I think (hopefully not too dangerously) that the solution here to is to implement more of it in C, given that the function is _already_ a mix of C and OCaml relying on subtle details of the C/OCaml boundaries:

```C
/* ... */

CAMLprim value caml_mutex_with_lock(value wrapper, value scope)
{
  CAMLparam2(wrapper, scope);
  CAMLlocal1(res);
  
  caml_mutex_lock(wrapper);

  res = caml_callback_exn(scope, Val_unit);
  if (Is_exception(result(res)) {
    res = Extract_exception(res);
    caml_mutex_unlock(wrapper);
    caml_raise(res);
  }

  caml_mutex_unlock(wrapper);

  CAMLreturn(res);
}
```

```ocaml
external with_lock : t -> scope:(unit -> 'a) -> 'a = "caml_mutex_with_lock"
```

I _think_ that that implementation still preserves the requirement that `with_lock` always unlocks the mutex regardless of aysnchronous exceptions.

**Comment by lpw25** [comment]
For some context here are a couple of examples that will need this attribute:

1. The [Nano_mutex](https://github.com/janestreet/core/blob/master/nano_mutex/src/nano_mutex.ml) library. The sections marked between `(* BEGIN ATOMIC *)` and `(* END ATOMIC *)` markers in that file will need to be lifted into functions and marked with `[@poll error]`. They all look like they should work with this PR.

2. The [`Ring` module withing the Jenga build system](https://github.com/janestreet/jenga/blob/v0.11/tenacious/lib/ring.ml). The blocks marked with `(* ATOMIC block *)` will need to become functions marked with `[@poll error]`. They look like they will mostly work -- although I think we will need to add a `external phys_equal : 'a -> 'a -> bool = "%eq"` line to the file to make it pass the check in dev builds.

**Comment by lpw25** [comment]
@gadmm If I understand your example correctly, it doesn't actually require atomicity for those section, it just requires protection from cancellation (i.e. asynchronous exceptions) in them. I think that your own masking proposals are probably a better solution than the `poll` attributes for those cases.

If you do really want to use a `poll` attribute for them, I think that rather than your `explicit` proposal it might be better to use `poll error` along with some other attribute to ignore specific function calls. Something like:

```ocaml
let[@poll error] with_lock m ~scope =
  let () = (lock m [@poll ignore]) in
  match (scope () [@poll ignore) with
  | x -> (unlock m [@poll ignore]); x
  | exception e -> (unlock m [@poll ignore]); raise e
```
where `[@poll ignore]` attached to a function call would prevent that call from being inlined and wouldn't raise the error if that call remained in the function.

**Comment by gadmm** [comment]
@dra27 Your remark raises the question of a target audience. Ideally `with_lock` in Coq should be written as `Fun.with_resource ~acquire:lock ~release:unlock` from my other PR, or equivalent, if this becomes possible one day, and `@poll explicit` should be avoided in a user program. `@poll error` can be safely handed to programmers, but is only useful in specialized application areas, so they would be told not to bother with it unless they have a good reason to use it. Programmers would be told something similar about `@poll explicit` but along with a warning, and be told to use instead high-level combinators such as `with_resource` whose support is the actual goal, except, again, in specialized application areas, for instance to implement similar high-level devices. So it is nice that the current proposal never causes deadlocks, but I think this should not be at the cost of preventing the implementation of solutions that cannot be written otherwise.

I see two problems with using a callback from C:
- The capture of call stacks of the form OCaml -> C -> OCaml by effect handlers is difficult to implement in multicore. If I read correctly <https://github.com/ocaml-multicore/ocaml-multicore/issues/146>, this is not yet supported, and if it is ever supported, it will come at the cost of instantiating a new C thread and doing some business with condition variables, that all callers of such a C function have to pay. This does not sound a good fit for `with_lock` and similar functions, which can be called frequently. This is different than calling C without calling back into OCaml.
- My goal is to enable writing resource-management combinators that can be more sophisticated than the well-bracketed (LIFO) combinators and notably allows to "move" resources. I doubt it is possible to write them with C calls, whose well-bracketing is hard-coded in the structure of the stack frames. These scope-bound resources are sometimes expressive enough, but when they are not, OCaml programmers fall back to manual resource management, which is error-prone and in need of a better solution (especially in the presence of interrupts, which in contrast are easy to support RAII-style... as long as there is some elementary control on polling locations for implementers such as outlined here).

@lpw25 The goal is indeed to support those masking proposals, and whatever comes after (which more realistically will first come in the form of libraries, whether as proof-of-concept or something usable, that would need to rely on such an attribute).

I like the idea of `@poll ignore` on function calls. It probably requires a similar amount of work as `@poll explicit` (as compared to `@poll error`, e.g. for flambda), but there remains the issue of inserted polls. The Safepoints PR hypothesised adding a polling location after `lock` returns in the future; in such event there should be a way to disable such polling location (using a similar attribute?). The latter is only a hypothetical example, but another concern is polling in function prologues. In native I believe they are removed by the dataflow analysis in practice, so this is likely to work but relies on implementation details. I fear it might make it harder or impossible to usefully extend the attribute to bytecode later on. The proposal to turn off polling locations is pragmatic in that it simplifies the specification and makes it simple to port to bytecode.

So I think we might be on the right track with an attribute on functions call, but this solution is not yet complete to me, and I am curious if you have further ideas.

Another example where I noticed we want to avoid polling at the start of a function is with the masking combinator from #8961 ([code](https://github.com/ocaml/ocaml/pull/8961/files#diff-facac154a3e5657ad3b1f7b8db538ed58f6aa4a2749dde974d97100f102aa725R129-R141)) (see the first `BEGIN ATOMIC` before the function argument, for lack of a better way to describe it, which implies that `mask` should not poll at the start). This is essentially the same problem as `with_resource`, modulo not inlining the masking combinator inside `with_resource`.

File: asmcomp/polling.ml (hunk 1 above)
Comment (sadiqj) [reply asmcomp/polling.ml:289]: Yes, you're right. Fix incoming.

**Comment by sadiqj** [comment]
Thanks for everyone's detailed input and examples.

I'm in agreement with @dra27 and @lpw25 here on @gadmm 's example. `scope` could do arbitrary things which may themselves poll (allocations, safepoints, etc..) and it seems it's possible to achieve the protection required with the existing C API.  There isn't anything in this implementation that would stop us supporting it with an additional annotation to allow for a pure OCaml implementation though.

The motivation behind this proposed annotation is that users are already having to manually verify that their atomic sections have no polling operations. This gets much harder when safepoints lands. The annotation just gives them a static guarantee it is correct and continues to be so if it compiles in the future.

**Comment by sadiqj** [comment]
Just trying to move this towards a full PR. It seems there at least motivating examples provided by @lpw25 on the JS codebases where this would provide additional safety.

@xavierleroy and @damiendoligez - would you happen to have views on the design?

If it's acceptable I can rebase atop the safepoints PR and we can start the work on flambda to get us to a full PR.

File: middle_end/flambda/flambda_utils.ml [hunk 4]
```diff
@@ -350,9 +350,14 @@ let make_closure_declaration
       ~body ~stub ~dbg:Debuginfo.none ~inline:Default_inline
       ~specialise:Default_specialise ~is_a_functor:false
       ~closure_origin:(Closure_origin.create (Closure_id.wrap id))
+      ~poll:Default_poll
   in
-  assert (Variable.Set.equal (Variable.Set.map subst free_variables)
+  begin
+    (* this is required because we hardcode poll behaviour above *)
+    assert( stub );
```
Comment (stedolan) [inline middle_end/flambda/flambda_utils.ml]: Should this `~stub` parameter just be removed?

Comment (sadiqj) [reply middle_end/flambda/flambda_utils.ml]: This is probably better than asserting. I'll remove it.

Comment (sadiqj) [reply middle_end/flambda/flambda_utils.ml]: Done. Also rebased.

**Review by stedolan — APPROVED** [review]

**Comment by sadiqj** [comment]
Just as an extra test, I extracted the circular ring buffer that @lpw25 mentioned earlier and added the annotations to the two functions with atomic sections. I had to add the external `phys_equal`.

Initially this fails with:
```File "ring.ml", line 1:
Error: Polling instructions in function annotated with [@poll error]
	function call at File "ring.ml", line 34, characters 2-5
	allocation at File "ring.ml", line 33, characters 13-45
```

Extracting the actual updating of the ring element (which is the bit that actually has the atomic section) to a separate function with the error annotation compiles: https://gist.github.com/sadiqj/0a8b68a89beec96758f39808f068f710

**Comment by mshinwell** [comment]
(This is still being reviewed, not yet ready for merge.)

**Comment by sadiqj** [comment]
typo failure is trunk's `Changes`

**Comment by mshinwell** [comment]
I think this is good to go (once CI passes).

**Review by mshinwell — APPROVED** [review]

**Comment by mshinwell** [comment]
Cherry-picked to 4.14, rev `3a2a1415dc9a62a742a6f631b8427a8804bae849`

File: lambda/translattribute.ml [hunk 5]
```diff
@@ -207,7 +245,7 @@ let add_inline_attribute expr loc attributes =
             (Warnings.Duplicated_attribute "inline")
       end;
       let attr = { attr with inline } in
-      check_local_inline loc attr;
+      check_poll_inline loc attr;
```
Comment (gasche) [inline lambda/translattribute.ml:248]: @sadiqj why was the `check_local_inline` call removed here, when it is kept below? I wonder if it's a small bug or an intentional change.

Comment (sadiqj) [reply lambda/translattribute.ml:248]: Good spot, I think this is a bug.

**Pull request merged**